### PR TITLE
Homework check

### PR DESCRIPTION
**Pull request by xakod**
Добавлен экран победы
Счетчик  ходов и таймер игры
Запись результатов идет в sqlite
Добавлена анимация передвижения плиток

В текущих реалиях я не совсем понял проверка правильности сгенерированного поля работает неверно, либо я так плохо собираю пятнашки. 
База данных требует скопировать себя в исполняемый каталог, если надо исправить, поправлю, но пока в планах не было.

### CODE REVIEW

File: GameBoard.qml [hunk 1]
```diff
@@ -4,11 +4,26 @@ import Game 1.0
 GridView {
     id: root
 
+    signal gameStateChanged(bool state)
+
+    function f(s) {
```
Comment (smay1613) [inline GameBoard.qml:9]: please use more understandable names for each entity

File: GameBoard.qml [hunk 2]
```diff
@@ -4,11 +4,26 @@ import Game 1.0
...
+        console.log("score is: ", s)
+    }
+
+    function setState(state) {
+        root.visible = state
```
Comment (smay1613) [inline GameBoard.qml:14]: adding ";" at the end of the line in functions is the good practice in qml

File: GameBoard.qml [hunk 3]
```diff
@@ -17,25 +32,28 @@ GridView {
...
+    GameController_qml {
         id: _gameController
     }
 
     Component.onCompleted: {
-        root.model =  _gameController.getModel();
+        root.model = _gameController.getModel()
+
+        root.gameStateChanged.connect(setState)
+        root.model.stateChanged.connect(gameStateChanged)
```
Comment (smay1613) [inline GameBoard.qml:57]: you can use Connections element

File: ToolBar.qml [hunk 4]
```diff
@@ -0,0 +1,55 @@
+import QtQuick 2.0
+import QtQuick.Controls 2.12
+
+Rectangle {
+    id: root
+
+    property int scoreValue
+    property alias newGame: _newGameButton
```
Comment (smay1613) [inline ToolBar.qml:8]: why it is used as alias?

File: ToolBar.qml [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    Button {
+        id: _newGameButton
+
+        anchors.left: parent.left
+        anchors.verticalCenter: parent.verticalCenter
+
+        text: "Start new game"
+    }
+    Button {
+        id: _secretButton
```
Comment (smay1613) [inline ToolBar.qml:30]: is it used?

File: VictoryScreen.qml [hunk 6]
```diff
@@ -0,0 +1,82 @@
+import QtQuick 2.0
+import QtQuick.Controls 2.12
+
+Rectangle {
+    id: root
+    signal restartClicked
+    signal gameFinished
+    signal gameStateChanged(bool state)
```
Comment (smay1613) [inline VictoryScreen.qml:8]: you can use State component for this purpose

File: VictoryScreen.qml [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    // color: "black"
+    border.color: "black"
+    border.width: 1
+    radius: 5
+
+    enabled: false
+    visible: false
+    Component.onCompleted: {
+        _newGameButton.clicked.connect(restartClicked)
+        root.gameFinished.connect(viewResult)
```
Comment (smay1613) [inline VictoryScreen.qml:21]: i think it looks more clear if you make a slot for the button, and call/emit something in it, also for the easier debugging purpose

File: VictoryScreen.qml [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+        font.pointSize: height / 4
+
+        height: root.height / 4
+        width: root.width / 4
+
+        anchors.left: root.left
+        anchors.right: root.right
+        anchors.top: root.top
+
+        // anchors.leftMargin: width / 10
```
Comment (smay1613) [inline VictoryScreen.qml:50]: always remove commented code, you aren't gonna need it

File: gameboard.cpp [hunk 9]
```diff
@@ -1,142 +1,228 @@
...
+  //           << ((inv & 1) ? "No Solution" : "Solution Exists");
+  return (inv % 2) == 0;
+  // return (inv & 1);
+}
 
-    return QVariant(static_cast<int>(m_raw_board[index_row].value));
+bool GameBoard::isBoardFinished() {
+  bool state{true};
+  size_t pos{0};
+  std::for_each(m_raw_board.begin(), m_raw_board.end(),
```
Comment (smay1613) [inline gameboard.cpp:115]: **std::is_sorted?**

File: gameboard.cpp [hunk 10]
```diff
@@ -1,142 +1,228 @@
...
+  // destination
+
+  if (beginMoveRows(QModelIndex(), index, index, QModelIndex(), destination)) {
+    if (!gameTimer.isValid())
+      gameTimer.start();
+
+    std::swap(m_raw_board.at(static_cast<size_t>(index)),
+              m_raw_board.at(static_cast<size_t>(distanceToHiddenElement)));
+    endMoveRows();
+    qDebug() << "Time is: " << gameTimer.elapsed()
```
Comment (smay1613) [inline gameboard.cpp:194]: Please move the timer to QML, imho it is not needed here

File: gameboard.cpp [hunk 11]
```diff
@@ -1,142 +1,228 @@
...
+             << " Score is: " << stepsCount();
+    setStepsCounter(++m_stepsCounter);
+
+    if (isBoardFinished()) {
+      if (gameTimer.isValid()) {
+        {
+          m_gameTime = static_cast<size_t>(gameTimer.elapsed());
+          database.writeResult(m_stepsCounter, m_gameTime);
```
Comment (smay1613) [inline gameboard.cpp:202]: move this logic to the controller

File: gameboard.cpp [hunk 12]
```diff
@@ -1,142 +1,228 @@
...
-    Position positionOfIndex {getRowCol(index)};
+    emit dataChanged(createIndex(0, 0),
+                     createIndex(static_cast<int>(m_boardsize), 0));
+  }
+
+  return true;
+}
 
-    auto hiddenElementIterator = std::find(m_raw_board.begin(), m_raw_board.end(), m_hiddenElementValue);
+int GameBoard::stepsCount() { return m_stepsCounter; }
```
Comment (smay1613) [inline gameboard.cpp:216]: move it to the controller

File: main.qml [hunk 13]
```diff
@@ -8,9 +8,45 @@ Window {
     height: 480
     title: qsTr("15-puzzle")
 
+    Component.onCompleted: {
+        _toolBar.restartClicked.connect(_gameBoard.newgame)
+        _toolBar.secretClicked.connect(_gameBoard.finish)
```
Comment (smay1613) [inline main.qml:13]: connect it in-place (just call in the slot)

File: sqlitemanager.cpp [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+  if (check_file.exists() && check_file.isFile())
+    qDebug() << "File exists";
+  else
+    qDebug() << "File doesnt exist: " << path;
+
+  return (check_file.exists() && check_file.isFile());
+}
+
+bool sqlitemanager::createDB(QSqlDatabase *base) {
+  auto *createTable = new QString(
```
Comment (smay1613) [inline sqlitemanager.cpp:44]: **why the dynamic memory is used?**

File: sqlitemanager.h [hunk 15]
```diff
@@ -0,0 +1,36 @@
+// Header for interaction with database
+// Warning!
+// Now you to copy default database file to location near your executable file
```
Comment (smay1613) [inline sqlitemanager.h:3]: **use QStandardPaths, as described in https://www.youtube.com/watch?v=rzlJq_njIDw**

File: sqlitemanager.h [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+// Please keep sure you did it
+
+#ifndef SQLITEMANAGER_H
```
Comment (smay1613) [inline sqlitemanager.h:6]: _pragma once is much better_

**Review by smay1613 — CHANGES_REQUESTED** [review]

File: GameBoard.qml (hunk 3 above)
Comment (xakod) [reply GameBoard.qml:57]: Maybe it's better. I saw only first lesson and docs, Connetions will be in the future

File: ToolBar.qml (hunk 5 above)
Comment (xakod) [reply ToolBar.qml:30]: It's secret (cheat) unvisible button for finish game in one click, i use it for test some functions

File: sqlitemanager.cpp (hunk 14 above)
Comment (xakod) [reply sqlitemanager.cpp:44]: This is my very old code from another project)

File: sqlitemanager.h (hunk 15 above)
Comment (xakod) [reply sqlitemanager.h:3]: Maybe it's time to go to second lesson and the to third) Thanx

File: sqlitemanager.h (hunk 16 above)
Comment (xakod) [reply sqlitemanager.h:6]: I read this https://stackoverflow.com/questions/1143936/pragma-once-vs-include-guards and i didnt find why include guards better or worse. One say this better, another say that better...

**Pull request closed without merging**